src/main/java/org/mockito/internal/util/KotlinInlineClassUtil.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2021 Mockito contributors
+ * This program is made available under the terms of the MIT License.
+ */
+package org.mockito.internal.util;
+
+import org.mockito.internal.stubbing.BaseStubbing;
+import org.mockito.internal.stubbing.OngoingStubbingImpl;
+import org.mockito.invocation.MatchableInvocation;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class KotlinInlineClassUtil {
+    // When mocking function, returning inline class, its return type is
+    // underlying type.
+    // So, `thenReturn` calls fails, because of non-compatible types.
+    public static boolean isInlineClassWithAssignableUnderlyingType(
+            Class<?> inlineClass, Class<?> underlyingType) {
+        try {
+            // All inline classes have 'box-impl' method, which accepts
+            // underlying type and returns inline class.
+            inlineClass.getDeclaredMethod("box-impl", underlyingType);
+            return true;
+        } catch (NoSuchMethodException e) {
+            return false;
+        }
+    }
+
+    private static Object unboxInlineClass(Object boxedValue)
+            throws InvocationTargetException, IllegalAccessException {
+        Class<?> inlineClass = boxedValue.getClass();
+        Method unboxImpl = null;
+        for (Method declaredMethod : inlineClass.getDeclaredMethods()) {
```

Comment:
The same would be required here, assuming that `unbox-impl` also has a pre-known method signature?